### PR TITLE
move log of pre-segmenter passes to `OptimizationPass::runPass`

### PR DESCRIPTION
**Pull request by liqiangxl**
**Issue:** code repeat when each pre-segmenter pass prints fusion in their own pass function
**Fix:** Similar to `GpuLower` passes, store all the pre-segmenter passes in a `std::pair<PassFunction, std::string /*pass name*/>`, then call each pass in order and print log after each pass.

**TODO:**
(1) If nothing is changed in a specific pass, we can simply print a msg saying  "fusion unchanged" instead of print the fusion again. 
(2) optimize the order of these pre-segmenter passes.

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build

**Comment by xwang233** [comment]
> The OOM failures seem strange since these cases are only request a small piece of memory. It looks like we are running multiple-processes on the same GPU. @xwang233
> 
> ```
> To reproduce: NVFUSER_TEST_RANDOM_SEED=1717081225 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='NVFuserTest.FusionAdvancedComputeAt3_CUDA'
> 00:16:36 C++ exception with description "CUDA out of memory. Tried to allocate 64.00 MiB. GPU 0 has a total capacity of 39.38 GiB of which 50.69 MiB is free. Process 441868 has 684.00 MiB memory in use. Process 445293 has 18.11 GiB memory in use. Process 446025 has 18.11 GiB memory in use. Process 447672 has 2.42 GiB memory in use.
> ```

I restarted the binary A100 job.

File: csrc/preseg_passes/pre_segmenter.cpp [hunk 1]
```diff
@@ -22,16 +23,35 @@ namespace nvfuser::preseg_passes {
...
+       "ExactMappedExtentSubstitutionPass"},
+      {OptimizationPass<AllocationDomainPass>::runPass, "AllocationDomainPass"},
+      {OptimizationPass<RemoveBcastSqueeze>::runPass, "RemoveBcastSqueeze"}};
+
+  bool is_log_enabled =
+      isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging);
+  if (is_log_enabled) {
+    debug() << "Fusion before PreSegmenter:" << std::endl;
+    fusion->printMath();
+  }
```
Comment (wujingyue) [inline csrc/preseg_passes/pre_segmenter.cpp:46]: Sorry, I didn't put too much thought on where exactly to log. 

How about moving this to after https://github.com/NVIDIA/Fuser/blob/c2dbfe4cf76fdce436d2899bf32fac0b2b1a0778/csrc/preseg_passes/optimization_pass.h#L53? The class OptimizationPass was created to handle common logics like this. Other examples include checking cycles and skipping disabled passes.

Comment (wujingyue) [reply csrc/preseg_passes/pre_segmenter.cpp:46]: That probably also requires to embed the tag/name in the pass. For example, each preseg pass will have a method `name()` that returns a string.

Comment (liqiangxl) [reply csrc/preseg_passes/pre_segmenter.cpp]: moved to after `DerivedClass::runPass`, added `name()` to each pass.
When run with `NVFUSER_DUMP=pre_segmenter_logging`, a typical output is
```
[ RUN      ] NVFuserTest.FusionRemoveEmptyMatmul_CUDA
Fusion before PreSegmenter optimizations:
Inputs:
  T0_g[ iS0{16}, iS1{0} ], __half
  T1_g[ iS2{0}, iS3{8} ], __half
Outputs:
  T5_g[ iS13{16}, iS14{8}, rS15{0} ], float

%kernel_math {
T2_l[ iS4{16}, bS5{1}, iS6{0} ]
   = broadcast( T0_g[ iS0{16}, iS1{0} ] )
T3_l[ bS7{1}, iS8{0}, iS9{8} ]
   = broadcast( T1_g[ iS2{0}, iS3{8} ] )
T4_l[ bS10{1}, iS12{8}, iS11{0} ]
   = Set.Permute( T3_l[ bS7{1}, iS8{0}, iS9{8} ], cache_op=Streaming )
T5_g[ iS13{16}, iS14{8}, rS15{0} ]
   = mma(T2_l[ iS4{16}, bS5{1}, iS6{0} ],
         T4_l[ bS10{1}, iS12{8}, iS11{0} ])
}

========================================
Fusion after pass: RemoveEmptyPass
Inputs:
  T0_g[ iS0{16}, iS1{0} ], __half
  T1_g[ iS2{0}, iS3{8} ], __half
Outputs:
  T6_g[ iS16{16}, iS17{8} ], float

%kernel_math {
T6_g[ iS16{16}, iS17{8} ]
   = full({16, 8}, float(0));
}

========================================
Fusion after pass: ConsecutiveCastPass
Inputs:
  T0_g[ iS0{16}, iS1{0} ], __half
  T1_g[ iS2{0}, iS3{8} ], __half
Outputs:
  T6_g[ iS16{16}, iS17{8} ], float

%kernel_math {
T6_g[ iS16{16}, iS17{8} ]
   = full({16, 8}, float(0));
}
```

File: csrc/preseg_passes/add_axioms.h [hunk 2]
```diff
@@ -16,6 +16,9 @@ class AddAxiomsPass : public OptimizationPass<AddAxiomsPass> {
 
  protected:
   NVF_API static void runPass(Fusion* fusion);
+  NVF_API static std::string name() {
```
Comment (wujingyue) [inline csrc/preseg_passes/add_axioms.h:19]: I understood you were copying the NVF_API from the other method. I doubt these NVF_APIs are needed, so let's try to remove them and make sure CI passes.

File: csrc/preseg_passes/mark_aliases_prepare.cpp [hunk 3]
```diff
@@ -107,8 +105,7 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
   }
 
   if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
-    debug() << "Fusion after MarkAliasesPreparePass:" << std::endl;
-    fusion->printMath();
+    debug() << "Fusion Transforms after MarkAliasesPreparePass:" << std::endl;
```
Comment (wujingyue) [inline csrc/preseg_passes/mark_aliases_prepare.cpp:108]: How about replacing MarkAliasesPreparePass with `name()` so we don't have two places hardcoding the name?

File: csrc/preseg_passes/pre_segmenter.cpp [hunk 4]
```diff
@@ -22,6 +25,11 @@ namespace nvfuser::preseg_passes {
 /*static*/ void PreSegmenter::runPass(Fusion* fusion) {
   FUSER_PERF_SCOPE("PreSegmenter::runPass");
 
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    debug() << "Fusion before PreSegmenter optimizations:" << std::endl;
```
Comment (wujingyue) [inline csrc/preseg_passes/pre_segmenter.cpp:29]: s/PreSegmenter/name()

**Review by wujingyue — APPROVED** [review]
Thanks -- looks great!

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
!build

File: csrc/preseg_passes/add_axioms.h (hunk 2 above)
Comment (liqiangxl) [reply csrc/preseg_passes/add_axioms.h]: You are right. It's a copy-paste. I'll remove & check CI.

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
Hi @xwang233 do you know how can I restart these two failed tests? `https://nv/e2E/95802234` I am not sure if they are real failures or some CI issues exist. Thanks.

**Comment by xwang233** [comment]
> Hi @xwang233 do you know how can I restart these two failed tests? `https://nv/e2E/95802234` I am not sure if they are real failures or some CI issues exist. Thanks.

Hi @liqiangxl , those are indeed network issue in CI. Sorry for the confusions. I just gave you access to restart those jobs.

**Pull request merged**